### PR TITLE
feat: type based version increments

### LINKED ISSUE #29: Allow more flexible version increments per commit type.
**Is your feature request related to a problem? Please describe.**
Gotagger is inflexible about how it tags various commit types. I would like the ability for it to ignore certain commit types (like "docs:" for instance) when calculating the repo tag.

**Describe the solution you'd like**
I would like Gotagger to take an optional configuration file that contains mappings of which commit types match with which version increments. The config file should have a default version increment for any commit types it doesn't recognize. If a config file is not provided, Gotagger should behave as it does now.

### PR DESCRIPTION
**Pull request by ninjamast3r**
- Users may provide a config file that maps commit types to version
increments. The default behavior is how gotagger behaved previously.
fixes #29

### CODE REVIEW

File: gotagger.go [hunk 1]
```diff
@@ -223,6 +230,47 @@ func (g *Gotagger) Version() (string, error) {
 	return versions[0], nil
 }
 
+// ReadCommitTypeMappings reads a config file containing mappings of commit type to semver increment. Mappings determine
+// how much to increment the semver based on the commit type. Unknown commit types will fall back to the config default.
+// Invalid increments will throw an error. Duplicate type definitions will take the last entry.
+func (g *Gotagger) ReadCommitTypeMappings() error {
```
Comment (ninjamast3r) [inline gotagger.go:236]: Move this to the config struct.

File: internal/commit/commit.go [hunk 2]
```diff
@@ -15,17 +16,97 @@ var (
 	footerRe = regexp.MustCompile(`^(?P<title>[-\w ]+): (?P<text>.*)`)
 )
 
-type Type string
+type IncrementConfig struct {
```
Comment (ninjamast3r) [inline internal/commit/commit.go:19]: Move this to public so we can use it in the config module

File: cmd/gotagger/main.go [hunk 3]
```diff
@@ -79,6 +80,7 @@ func (g *GoTagger) Run() int {
 	flags.BoolVar(&g.tagRelease, "release", g.boolEnv("release", false), "tag HEAD with the current version if it is a release commit")
 	flags.StringVar(&g.versionPrefix, "prefix", g.stringEnv("prefix", "v"), "set a prefix for versions")
 	flags.StringVar(&g.dirtyIncrement, "dirty", g.stringEnv("dirty", ""), "how to increment the version for a dirty checkout [minor, patch]")
+	flags.StringVar(&g.commitMapfile, "commit-map-file", g.stringEnv("commit_map_file", ""), "path to a custom configuration for how to increment the version based on commit type")
```
Comment (ninjamast3r) [inline cmd/gotagger/main.go:83]: make this just `config` and `gotagger.json`.

File: internal/commit/commit.go [hunk 4]
```diff
@@ -15,17 +16,97 @@ var (
...
+	TypeBuild       = "build"
+	TypeChore       = "chore"
+	TypeCI          = "ci"
+	TypeDocs        = "docs"
+	TypeRevert      = "revert"
+)
+
+var defaultCommitTypeMapper = map[string]Increment{
+	TypeFeature:     IncrementMinor,
+	TypeRelease:     IncrementPatch,
```
Comment (ninjamast3r) [inline internal/commit/commit.go:64]: Patch can just be the default for this. Don't need the specify the rest.

File: README.md [hunk 5]
```diff
@@ -19,41 +20,32 @@
...
-between the current commit and the latest tag,
-to determine if the most significant change was a
-feature,
-bug fix,
-or breaking change
-per the [Conventional Commits] format.
-`gotagger` then increments the base version accordingly
-and print the new version.
+`gotagger` is a library and CLI for tagging releases in git repositories as part
+of a CI process.
```
Comment (wfscheper) [inline README.md:24]: You're going to need to re-format all of this. See https://github.com/sassoftware/gotagger/blob/main/CONTRIBUTING.md#write-documentation

File: README.md [hunk 6]
```diff
@@ -77,53 +68,73 @@ build/$(go env GOOS)/gotagger
...
+In some cases, you may wish to change which semantic version field should
+increment based on which commit type is provided. You can do this via
+a [config file](gotagger.json). The config file contains a mapping of commit
+type to semver increment and a default increment to use if it encounters an
+unknown commit type. For example:
+
+```json
+{
+  "incrementMappings": {
+    "release": "major",
```
Comment (wfscheper) [inline README.md:104]: "release" means something specific to gotagger, so I don't think we should allow it in the mapping. "release" must at least be "patch", or else gotagger will try to create the same tag twice (and fail) if you try to do two release commits back-to-back. If "release" is "minor" or "major", then you will never be able to release any less significant version increment.

Also the conventional commit spec is explicit about how to increment the major version, so should gotagger even let you configure a "major" increment?

File: README.md [hunk 7]
```diff
@@ -77,53 +68,73 @@ build/$(go env GOOS)/gotagger
...
+
+Running:
+
+```bash
+gotagger -config ./gotagger.json
+```
+
+If no configuration is provided, Gotagger defaults to the current functionality,
+which is equivalent to what is defined in
+[gotagger.json](gotagger.json)
```
Comment (wfscheper) [inline README.md:120]: Should gotagger default to reading a "gotagger.json" config file in the current directory if it exists, or is the explicit flag required?

File: cmd/gotagger/main.go [hunk 8]
```diff
@@ -61,6 +61,7 @@ type GoTagger struct {
 	tagRelease     bool
 	versionPrefix  string
 	dirtyIncrement string
+	commitMapfile  string
```
Comment (wfscheper) [inline cmd/gotagger/main.go:64]: Let's name this `configFile`. Pretty much guaranteed it'll hold more than just the commit mapping...

File: gotagger.go [hunk 9]
```diff
@@ -65,6 +66,12 @@ type Config struct {
 	// if there are no new commits, but the worktree is "dirty".
 	DirtyWorktreeIncrement string
 
+	// IncrementConfigFile path to the file where the increment configs are defined.
+	IncrementConfigFile string
```
Comment (wfscheper) [inline gotagger.go:70]: Let's just go ahead and bite the config file bullet: add json tags to the Config struct and Unmarshal it from the config file. I'll take a stab at that...

**Review by wfscheper — CHANGES_REQUESTED** [review]

File: cmd/gotagger/main.go (hunk 8 above)
Comment (ninjamast3r) [reply cmd/gotagger/main.go]: Done

File: README.md (hunk 7 above)
Comment (ninjamast3r) [reply README.md]: If gotagger.json is there, it'll use it. That behavior is overridden by the config file. If it doesn't exist, it'll take the default.

File: README.md (hunk 6 above)
Comment (ninjamast3r) [reply README.md]: I changed it so release and major increments can't be defined in the increment config.

File: README.md (hunk 5 above)
Comment (ninjamast3r) [reply README.md]: Done

File: gotagger.go [hunk 10]
```diff
@@ -75,6 +82,55 @@ type Config struct {
 	*/
 }
 
+// ReadCommitTypeMappings reads a config file containing mappings of commit type to semver increment. Mappings determine
+// how much to increment the semver based on the commit type. Unknown commit types will fall back to the config default.
+// Invalid increments will throw an error. Duplicate type definitions will take the last entry.
+func (c *Config) ReadCommitTypeMappings() error {
```
Comment (wfscheper) [inline gotagger.go:88]: Let's refactor this to `ParseJSON([]byte) error`, and drop the ConfigFile member.

Embed the fields from IncrementConfig in Config.

**Review by wfscheper — CHANGES_REQUESTED** [review]

File: gotagger.go (hunk 10 above)
Comment (ninjamast3r) [reply gotagger.go]: Done

**Review by wfscheper — APPROVED** [review]

**Review by wfscheper — CHANGES_REQUESTED** [review]
Drat, I forgot you need to add a fragment file. Name it `.stentor.d/30.feat.confit-patch-increment.md`, and just put in a quick summary of the new feature. Here's an example: https://raw.githubusercontent.com/sassoftware/gotagger/main/.stentor.d/24.feat.add-dirty-option.md

File: cmd/gotagger/main.go [hunk 11]
```diff
@@ -79,6 +80,17 @@ func (g *GoTagger) Run() int {
...
 	flags.StringVar(&g.dirtyIncrement, "dirty", g.stringEnv("dirty", ""), "how to increment the version for a dirty checkout [minor, patch]")
+	flags.StringVar(&g.configFile, "config", g.stringEnv("config", ""), "path to the gotagger configuration file.")
+
+	if g.configFile == "" {
+		// If there's no config file provided, check for one locally.
+		defaultConfig := filepath.Join(g.WorkingDir, "gotagger.json")
+		_, err := os.Stat(defaultConfig)
+		if err == nil {
+			g.configFile = defaultConfig
+			log.Println("using config file " + defaultConfig)
```
Comment (wfscheper) [inline cmd/gotagger/main.go:91]: Delete this, it always outputs and gotagger's only output should be the version.

**Review by wfscheper — CHANGES_REQUESTED** [review]

**Comment by wfscheper** [comment]
Trying to figure out why golangci-lint is giving us problems again...

**Comment by wfscheper** [comment]
Okay, think it's fixed. You'll need to rebase again, though.

**Review by wfscheper — APPROVED** [review]

**Pull request merged**